### PR TITLE
Directly piping files into external commands

### PR DESCRIPTION
**Pull request by bptato**
Includes the following:

* PIPE_LINK: download the link pointed to by the current anchor, and pipe
  it to an external command.

* x-nosaveinput mailcap extension field: avoid creating temporary files for
  a given input.

* README.mailcap English translation.

Use case for x-nosaveinput: mainly downloading very large files, e.g. FLAC
audio or video files. I've been using x-nosaveinput for a while to view
videos/music on my file system and through ftp.

But maybe it's useful for other things too, e.g. a markdown to html converter
could operate directly on the input stream instead of having to wait for
the entire file to be downloaded.

On the other hand, I am not exactly sure about the utility of PIPE_LINK.
I originally added it for the same purpose, but I haven't had much use for
it ever since I implemented x-nosaveinput. For now I have included it,
but if requested I can remove it from the PR.

On docs: I have added a description of x-nosaveinput to the Japanese mailcap,
and translated it into English so it can be read by more people. Corrections
are welcome, as my Japanese is not that good yet.

### CODE REVIEW

File: file.c [hunk 1]
```diff
@@ -8236,7 +8249,8 @@ save2tmp(URLFile uf, char *tmpf)
 
 	buf = NewWithoutGC_N(char, SAVE_BUF_SIZE);
 	while ((count = ISread_n(uf.stream, buf, SAVE_BUF_SIZE)) > 0) {
-	    if (fwrite(buf, 1, count, ff) != count) {
+	    int c = fwrite(buf, 1, count, ff);
+	    if (c != count) {
```
Comment (rkta) [inline file.c]: Why this change?

Comment (bptato) [reply file.c]: My guess is that I was going to use c again, but then I didn't and forgot about the line :/
I will change it back; thank you for your input.

**Comment by rkta** [comment]
On Sat, Jul 29, 2023 at 06:57:36AM -0700, bptato wrote:
> Includes the following:
[...]

> On the other hand, I am not exactly sure about the utility of PIPE_LINK.
> I originally added it for the same purpose, but I haven&#39;t had much use for
> it ever since I implemented x-nosaveinput. For now I have included it,
> but if requested I can remove it from the PR.

I finally had time to test the PIPE_LINK command and I have to say I'm
not convinced.

On the one hand, I already thought about a curl like functionality to
pipe links into some tool, e.g. to apply a patch or to avoid downloading
a tar archive.

My main gripe is the lack of feedback if the call succeeded. My first
try was to pipe something to less and just nothing happened. I had to
look into the code to understand that I was expecting the wrong things
here. Although that could be fixed by better documentation I don't think
having absolutely no way to see any errors makes a good UX.

> On docs: I have added a description of x-nosaveinput to the Japanese mailcap,
> and translated it into English so it can be read by more people. Corrections
> are welcome, as my Japanese is not that good yet.

Thanks, great to have it in English - my Japanese is non-existing ;-)

File: file.c [hunk 2]
```diff
@@ -8286,19 +8284,28 @@ doExternal(URLFile uf, char *type, Buffer *defaultbuf)
...
 
-    if (mcap->nametemplate) {
-	tmpf = unquote_mailcap(mcap->nametemplate, NULL, "", NULL, NULL);
-	if (tmpf->ptr[0] == '.')
-	    ext = tmpf->ptr;
+    if ((mcap->flags & MAILCAP_NOSAVEINPUT)) {
+	if (uf.scheme == SCM_LOCAL) {
+	    tmpf = Strnew_charp(uf.url);
+	} else {
+	    tmpf = Strnew_charp("-"); /* make it a pipe */
```
Comment (rkta) [inline file.c:8299]: IIUC, this uses '-' as file name for the command. This assumes every command understands '-' to mean "Read from stdin".  It this standardized? Could we do it w/o this assumption and let the user add it if it's needed to make the command work?

**Comment by bptato** [comment]
I no longer think this is the best way to do what I wanted.

On x-nosaveinput:
The idea is not bad, but the implementation is horrible: it is a proprietary
extension incompatible with programs that do the right thing.

In fact, we can achieve streaming into external programs by just directly piping
the input file into the external command in all cases where no template string
(%s) has been specified.

So in the proposed system:
```
# This is piped.
image/*; feh -
# This is downloaded before the entry is opened.
image/*; feh %s
# This is piped as well, then the output is redirected into an HTML buffer.
text/markdown; pandoc; x-htmloutput
# This is downloaded before the pandoc call, then the output is read in as HTML.
text/markdown; pandoc %s; x-htmloutput
```

Another use case for x-nosaveinput though, is that it makes it possible to
stream downloaded data *and* pass the file path. Luckily there is also a way
to support this functionality without backwards-incompatible extensions:
we can simply set an environment variable to the target URL before the mailcap
entry is executed.

On PIPE_LINK:
Probably should be dropped. For my use case it has been superseded by mailcap,
and to make it actually useful it would have to be fundamentally re-designed
(i.e. it would have to support the same parameters mailcap does).

I intend to prepare a different PR where I implement this.
